[DOCS-1131] Document run metrics automations (#1209)
wandb/docs: content/guides/core/automations/automation-events.md

{{< readfile file="/_includes/enterprise-cloud-only.md" >}}
{{% /pageinfo %}}

> An automation can start when a specific event occurs within a project's or registry. This page describes the events that can trigger an automation within each scope.

Reviewer: ```suggestion
An automation can start when a specific event occurs within a project or registry. This page describes the events that can trigger an automation within each scope.
```

- The project and all of its collections (the default).
- All collections that match the filter you supply.

### Artifact events
This section describes the events related to an artifact that can trigger an automation.

- **Linking a new artifact**: Test and validate new models, datasets or dataset automatically.
- **Creating a new version of an artifact**: Apply recurring actions to each version of an artifact. For example, start a training job when a new dataset artifact version is created.
- **Adding a new alias to a version of an artifact**: Trigger a special step of your workflow when a new artifact version in a project or collection has a specific label or alias applied. For example, run a series of downstream processing steps when an artifact has the `test-set-quality-check` alias applied.

### Run metrics events
This section describes the run metrics events that can trigger an automation.

> - **Run metrics threshold met**: Trigger a workflow when, for a given metric, a run or the average of a number of runs meets the threshold you specify for a given metric.

Reviewer: ```suggestion
- **Run metrics threshold met**: Trigger a workflow when, for a given metric, the average of the metric value meets the threshold you specify for a given metric.
```

It's not across runs, it's all within a single run.

Author: Done

> - **Run metrics change threshold met**: Trigger a workflow when, for a given metric, a run or the average of a number of runs increases or decreases by the threshold you specify for a given metric.

Reviewer: ditto above
```suggestion
- **Run metrics change threshold met**: Trigger a workflow when, for a given metric, the average of the metric value increases or decreases by the threshold you specify for a given metric.
```

Author: Done

To set up the metric for a run metrics event, you specify:

- An optional run name filter. Only runs matching this filter can trigger the automation.

> - The metric.

Reviewer: it's probably worth discussing system metrics and how to get the names of those

Author: Done

- The threshold for evaluation.
- The comparison to make. Choices differ for each type of event. Refer to [Run metrics comparison choices]({{< relref "#run-metrics-comparison-choices" >}}).

#### Run metrics comparison choices

> When configuring an automation for a run metric event, you can configure how to compare the run metric value with the threshold you specify. Your choices depend on the event type.

Reviewer: So this section has a tricky job: it needs to explain how the change threshold metric works. This isn't that complicated, but there are a lot of ways that this *could* be done, and we need to make sure we're very explicit about how it works so that users 

I think it's worth separating out the two types of run metrics (threshold vs change threshold) and explaining the algorithms separately, specifically what values are included in the windows (eg that the current & prior don't overlap and relative ordering), and how the windows are compared (for the change threshold)

Let me know if you'd like to meet to discuss the details of this.

Author: I think this is done, PTAL!

For **Run metrics threshold met** events, you can configure:
1. The number of logged values to average across (defaults to 5).

> 1. How to compare the values with the threshold:

Reviewer: Do we really need to list out the values in the options dropdown? I don't see what users will learn from this. if we needed to explain what they did, we should do that, but I don't think we need to

Author: Done

- All collections in the project that match the filter you supply.

### Artifact events
An automation can watch for these [artifact]({{< relref "/guides/core/artifacts/" >}}) events:
- **A new version of an artifact is created in a collection**: Apply recurring actions to each version of an artifact. Specifying a collection is optional. For example, start a training job when a new dataset artifact version is created.
- **An artifact alias is added**: Trigger a special step of your workflow when a new artifact version in a project or collection has a specific alias applied. For example, run a series of downstream processing steps when an artifact has the `test-set-quality-check` alias applied.

### Run metrics events
An automation can watch for changes in a run metric:

- **Run metrics threshold met**: Trigger a workflow when for a given metric, a single logged value or the average logged values meets the threshold you specify.
- **Run metrics change threshold met**: Trigger a workflow when for a given metric, a single logged value or the average logged value changes by the threshold you specify.

{{% alert %}}

> You can create aautomations triggered by system metrics such as the `cpu` metric, which tracks the percentage of CPU utilization. W&B logs system metrics automatically every 15 seconds. Find each metric's name and details in [System metrics]({{< relref "/guides/models/app/settings-page/system-metrics.md" >}}).

Reviewer: ```suggestion
You can create automations triggered by system metrics such as the `cpu` metric, which tracks the percentage of CPU utilization. W&B logs system metrics automatically every 15 seconds. Find each metric's name and details in [System metrics]({{< relref "/guides/models/app/settings-page/system-metrics.md" >}}).
```

{{% /alert %}}

For a run metric event, you can configure how to compare the run metric value with the threshold you specify. Your choices depend on the event type.

#### Threshold
For **Run metrics threshold met** events, you configure:
1. The number of logged values to average across (defaults to 5).
1. How to compare the values with the threshold.

For example, trigger an automation when `accuracy` exceeds `.6`.

#### Change threshold

For **Run metrics change threshold met** events, you can configure:
1. The current window of logged values to average across (defaults to the 10 most recent logged values).
1. The prior window of logged values to average across (defaults to the 50 most recent logged values prior to the current window).
1. Whether to evaluate the values as relative or absolute (defaults to **Relative**).
1. How to compare the values with the threshold:
      - Increases by at least
      - Decreases by at least
      - Increases or decreases by at least

> The current and prior window do not overlap. By default, the current window includes the 10 most recent logged values, and the prior window includes the 50 logged values prior to the current window.

Reviewer: also @mdlinville I realized something I'm working on could affect these docs - not sure if you wanted to change them at all 

https://app.graphite.dev/github/pr/wandb/core/28397/feat(artifacts)-Simplified-alert-creation-UI-for-workspaces is the PR

basically, just in workspaces we have a simplified UI for creating slack notifications when you're making automations on a metric.

Author: I'm having trouble accessing that. If it is not already available, we should address it as a fast follow. These docs are intended to be published very very soon.

- Collection level: All collections in the project that match the filter you specify.

### Artifact events
This section describes the events related to an artifact that can trigger an automation.

- **Linking a new artifact**: Test and validate new models, datasets or dataset automatically.
- **Creating a new version of an artifact**: Apply recurring actions to each version of an artifact. For example, start a training job when a new dataset artifact version is created.
- **Adding a new alias to a version of an artifact**: Trigger a special step of your workflow when a new artifact version in a project or collection has a specific label or alias applied. For example, run a series of downstream processing steps when an artifact has the `test-set-quality-check` alias applied.

### Run metrics events
An automation can watch for changes in a run metric:

- **Run metrics threshold met**: Trigger a workflow when for a given metric, a single logged value or the average logged values meets the threshold you specify.
- **Run metrics change threshold met**: Trigger a workflow when for a given metric, a single logged value or the average logged value changes by the threshold you specify.

> Each run is considered separately, even if the automation has a run filter that matches with multiple runs. Each run's values are put into a separate window and compared to the threshold separately. If an automation starts due to the result of one run, it can still start for another matching run. A particular run metric automation runs only once per run per 24-hour period.

Reviewer: It feels a bit random having this here - we have sections for threshold, does it make sense to add a section for the run filter, and put this in it?

This would be fine to do in a follow on change since I know you're under time pressure.

Author: Good idea, PTAL!

{{% alert %}}

> You can create automations triggered by system metrics such as the `cpu` metric, which tracks the percentage of CPU utilization. W&B logs system metrics automatically every 15 seconds. Find each metric's name and details in [System metrics]({{< relref "/guides/models/app/settings-page/system-metrics.md" >}}).

Reviewer: ditto that this feels random put in here - I don't see why it needs to have an alert. If we created a 'metric name' section, we could put this in there

This would be fine to do in a follow on change since I know you're under time pressure.

Author: I made an attempt -- PTAL!

- Project level: The automation watches for the event taking place on any collection in the project.
- Collection level: All collections in the project that match the filter you specify.

### Artifact events
This section describes the events related to an artifact that can trigger an automation.

> - **Linking a new artifact**: Test and validate new models, datasets or dataset automatically.

Reviewer 2: ```suggestion
- **Linking a new artifact**: Test and validate new models or datasets automatically.
```

- **Creating a new version of an artifact**: Apply recurring actions to each version of an artifact. For example, start a training job when a new dataset artifact version is created.

> - **Adding a new alias to a version of an artifact**: Trigger a special step of your workflow when a new artifact version in a project or collection has a specific label or alias applied. For example, run a series of downstream processing steps when an artifact has the `test-set-quality-check` alias applied.

Reviewer 2: ```suggestion
- **Adding a new alias to a version of an artifact**: Trigger a specific step of your workflow when a new artifact version in a project or collection has a specific label or alias applied. For example, run a series of downstream processing steps when an artifact has the `test-set-quality-check` alias applied.
```

- A metric in a run's history.
- A [system metric]({{< relref "/guides/models/app/settings-page/system-metrics.md" >}}) such as `cpu`, which tracks the percentage of CPU utilization. W&B logs system metrics automatically every 15 seconds.

#### Threshold
For **Run metrics threshold met** events, you configure:
1. The number of logged values to average across (defaults to 5).
1. How to compare the values with the threshold.

For example, trigger an automation when `accuracy` exceeds `.6`.

#### Change threshold
For **Run metrics change threshold met** events, you can configure:
1. The current window of logged values to average across (defaults to the 10 most recent logged values).
1. The prior window of logged values to average across (defaults to the 50 most recent logged values prior to the current window).
1. Whether to evaluate the values as relative or absolute (defaults to **Relative**).
1. How to compare the values with the threshold:
      - Increases by at least
      - Decreases by at least
      - Increases or decreases by at least

The current and prior window do not overlap.

#### Run filters
This section describes how the automation selects runs to evaluate.

> - Each run is considered separately, and the automation can potentially start due to the result of multiple runs.

Author: ```suggestion
- Each run is considered separately and can potentially trigger the automation.
```

> - Each run's values are put into a separate window and compared to the threshold separately.

Reviewer 2: What is meant by "windows"? Are windows like bins when it comes to histograms?

Author: Not quite! Because this is about an amount of change, there is a window for "current" runs and a window for "historical" runs. The two windows run consecutively but don't overlap. By default, the most recent 10 runs are averaged and compared with the next most recent 50 runs. Do I have it right, @ssisk ?

Reviewer 2: Ah. I see. Perhaps had a quick, short sentence that summarizes how windows work (and why it's important) before listing the window options (i.e. before the list)

Reviewer: yes, matt's description is correct

Author: Restructured a bit in the latest iteration, thank you both for helping me get this right.

- If you specify a run filter, only matching runs can start the automation.

> - A particular run metric automation runs at maximum once per run per 24-hour period.

Author: ```suggestion
- In a 24 hour period, a particular automation can fire at most once per run.
```

The current and prior window do not overlap. By default, the current window includes the 10 most recent logged values, and the prior window includes the 50 logged values prior to the current window.

#### Run filters
This section describes how the automation selects runs to evaluate.

> - Each run is considered separately, and the automation can potentially start due to the result of multiple runs.

Reviewer: I find this sentence a bit confusing and a bit repetitive with lines 91 and 92 - I wonder if we can just drop this line and move line 92 first, maybe tweaking it to say something like "if you specify a run filter, all matching run will be evaluated by the automation. If no run filter is specified all runs in a project will be evaluated."

Author: PTAL

Reviewer: thanks, much clearer! I assume that you're thinking that users will already be assuming that run metric automations without a run filter will match with all runs in the project?

Author: Nope, I would have missed this. Addressed.

- Each run's values are put into a separate window and compared to the threshold separately.
- If you specify a run filter, only matching runs can start the automation.

> - A particular run metric automation runs at maximum once per run per 24-hour period.

Reviewer: I think this is confusing because `run` is ambiguous, and eg could imply "a particular run metric can only evaluate metric values once per day".  We use the terms "fire" or "trigger" to refer to what happens when we execute the action that the user  has specified to avoid the confusion.

Author: PTAL to see if this is better.
